File: backend/app/schemas/semantic_index.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SemanticIndexCreateRequest(BaseModel):
    meaning: str = Field(min_length=1)
    summary: str | None = None
    embedding_text: str | None = None
    entry_node_ids: list[str] = Field(min_length=1)
    metadata: dict[str, Any] | None = None


class SemanticIndexResponse(BaseModel):
    id: str
    meaning: str
    summary: str | None
    embedding_text: str | None
    entry_node_ids: list[str]
    vector_status: str
    metadata: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime


class SemanticIndexEntryPointsResponse(BaseModel):
    entry_node_ids: list[str]


class SemanticIndexSearchResult(BaseModel):
    id: str
    meaning: str
    summary: str | None
    embedding_text: str | None
    entry_node_ids: list[str]
    vector_status: str
    metadata: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime
    score: float


class SemanticIndexSearchResponse(BaseModel):
    results: list[SemanticIndexSearchResult]

File: frontend/src/types/semanticIndex.ts
/**
 * Mirrors backend `app/schemas/semantic_index.py`.
 * Search results match `SemanticIndexSearchResult`; list wrapper matches artifact list style.
 */

export interface SemanticIndexResponse {
  id: string;
  meaning: string;
  summary: string | null;
  embedding_text: string | null;
  entry_node_ids: string[];
  vector_status: string;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface SemanticIndexListResponse {
  items: SemanticIndexResponse[];
}

export interface SemanticIndexEntryPointsResponse {
  entry_node_ids: string[];
}

export interface SemanticIndexSearchResult extends SemanticIndexResponse {
  score: number;
}

export interface SemanticIndexSearchResponse {
  results: SemanticIndexSearchResult[];
}
